File: backend/app/services/ai/insight_response.py
"""Build v2 AIInsightResponse from parsed AI JSON + payload.

Pure function — no LLM, no DB.

Exports:
    build_insight_response(ai_json, payload, prev) -> dict

Return shape matches SPEC §4.5 (TypeScript AIInsightResponse):
    { symbol, updatedAt, header, briefing, layers:{L1..L5}, rawInput }
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from app.services.ai.insight_fragments import parse_fragments
from app.services.ai.insight_levels import briefing_variant, status_level

# ── Constants ────────────────────────────────────────────────────────────────

_LAYER_NAMES: dict[str, str] = {
    "L1": "Xu hướng",
    "L2": "Thanh khoản",
    "L3": "Dòng tiền",
    "L4": "Nội bộ",
    "L5": "Tin tức",
}

_RECOMMENDATION_SET = frozenset(
    {"Chờ điểm mua", "Có thể mua thử", "Quan sát thêm", "Nên giảm bớt", "Bán bớt"}
)

# Sentinel phrases that mean "no change" in the diff field.
_NO_CHANGE_SENTINELS = (
    "ổn định",
    "lần đầu",
)


# ── Helpers ──────────────────────────────────────────────────────────────────


def _format_volume(vol: Any) -> str:
    """Format a raw volume number to a short string like '18.0M' or '350K'."""
    try:
        v = float(vol)
    except (TypeError, ValueError):
        return ""
    if v >= 1_000_000:
        return f"{v / 1_000_000:.1f}M"
    if v >= 1_000:
        return f"{v / 1_000:.1f}K"
    return str(int(v))


def _build_header(payload: dict[str, Any]) -> dict[str, Any]:
    """Extract StockHeader from payload (price_board + company_overview/details)."""
    sym = str(payload.get("symbol", "")).upper()

    price_board = payload.get("price_board")
    pb: dict[str, Any] = {}
    if isinstance(price_board, list) and price_board:
        pb = price_board[0] if isinstance(price_board[0], dict) else {}

    # Price fields — key names produced by vietcap.fetch_price_board
    price = pb.get("close_price") or pb.get("matchedPrice") or pb.get("close") or 0
    ref = pb.get("reference_price") or pb.get("refPrice") or pb.get("referencePrice") or 0
    high = pb.get("high_price") or pb.get("high") or pb.get("highPrice") or 0
    low = pb.get("low_price") or pb.get("low") or pb.get("lowPrice") or 0
    total_vol = pb.get("total_volume") or pb.get("totalVolume") or pb.get("nmTotalTradedQty") or 0
    exchange = pb.get("exchange") or ""

    # Change percent
    change_pct: float = 0.0
    if ref and price:
        try:
            change_pct = round((float(price) - float(ref)) / float(ref) * 100, 2)
        except (TypeError, ValueError, ZeroDivisionError):
            change_pct = 0.0

    # Sector / index group from company_details (Vietcap fields)
    company_details = payload.get("company_details") or {}
    company_overview = payload.get("company_overview") or {}

    sector = (
        company_details.get("icb_name_2")
        or company_details.get("icb_name_3")
        or company_details.get("industryName")
        or ""
    )
    index_group = (
        company_details.get("stock_type")
        or company_details.get("index_type")
        or company_details.get("indexGroup")
        or company_overview.get("exchange")
        or exchange
        or ""
    )

    # isLive: True if price data present (not None/0)
    is_live = bool(pb)

    return {
        "symbol": sym,
        "sector": sector,
        "indexGroup": index_group,
        "price": price,
        "changePercent": change_pct,
        "high": high,
        "low": low,
        "volume": _format_volume(total_vol),
        "isLive": is_live,
    }


def _diff_has_change(diff_text: str) -> bool:
    """Return True if the diff is not a sentinel 'no-change' phrase."""
    lower = diff_text.lower()
    return not any(sentinel in lower for sentinel in _NO_CHANGE_SENTINELS)


def _build_diff(diff_text: str, *, is_first: bool) -> dict[str, Any]:
    return {
        "text": parse_fragments(diff_text),
        "hasChange": _diff_has_change(diff_text),
        "isFirstAnalysis": is_first,
    }


# ── Layer builders ───────────────────────────────────────────────────────────

def _layer_fields_l1(layer: dict[str, Any]) -> list[dict[str, Any]]:
    fields = []
    if layer.get("xu_huong"):
        fields.append({"label": "Xu hướng", "value": parse_fragments(str(layer["xu_huong"]))})
    if layer.get("statusLabel"):
        fields.append({"label": "Trạng thái", "value": parse_fragments(str(layer["statusLabel"]))})
    if layer.get("ho_tro"):
        fields.append({"label": "Hỗ trợ", "value": parse_fragments(str(layer["ho_tro"]))})
    if layer.get("khang_cu"):
        fields.append({"label": "Kháng cự", "value": parse_fragments(str(layer["khang_cu"]))})
    if layer.get("da_gia"):
        fields.append({"label": "Đà giá", "value": parse_fragments(str(layer["da_gia"]))})
    return fields


def _layer_fields_l2(layer: dict[str, Any]) -> list[dict[str, Any]]:
    fields = []
    if layer.get("thanh_khoan"):
        fields.append({"label": "Thanh khoản", "value": parse_fragments(str(layer["thanh_khoan"]))})
    if layer.get("cung_cau"):
        fields.append({"label": "Cung–Cầu", "value": parse_fragments(str(layer["cung_cau"]))})
    if layer.get("tac_dong"):
        fields.append({"label": "Tác động", "value": parse_fragments(str(layer["tac_dong"]))})
    return fields


def _layer_fields_l3(layer: dict[str, Any]) -> list[dict[str, Any]]:
    fields = []
    if layer.get("khoi_ngoai"):
        fields.append({"label": "Khối ngoại", "value": parse_fragments(str(layer["khoi_ngoai"]))})
    if layer.get("tu_doanh"):
        fields.append({"label": "Tự doanh", "value": parse_fragments(str(layer["tu_doanh"]))})
    if layer.get("statusLabel"):
        fields.append({"label": "Tác động", "value": parse_fragments(str(layer["statusLabel"]))})
    return fields


def _layer_fields_l4(layer: dict[str, Any]) -> list[dict[str, Any]]:
    fields = []
    if layer.get("noi_bo"):
        fields.append({"label": "Nội bộ", "value": parse_fragments(str(layer["noi_bo"]))})
    if layer.get("khoi_luong_tong"):
        fields.append({"label": "Khối lượng tổng", "value": parse_fragments(str(layer["khoi_luong_tong"]))})
    if layer.get("statusLabel"):
        fields.append({"label": "Tác động", "value": parse_fragments(str(layer["statusLabel"]))})
    return fields


def _layer_fields_l5(layer: dict[str, Any]) -> list[dict[str, Any]]:
    fields = []
    if layer.get("tong_quan"):
        fields.append({"label": "Tổng quan", "value": parse_fragments(str(layer["tong_quan"]))})
    # Tin material — render as a list of items with title + tag + tac_dong_ngan
    tin_material = layer.get("tin_material") or []
    if tin_material:
        material_lines = "; ".join(
            f"{item.get('tieu_de', '')} [{item.get('tag', '')}] — {item.get('tac_dong_ngan', '')}"
            for item in tin_material
            if isinstance(item, dict)
        )
        fields.append({"label": "Tin material", "value": parse_fragments(material_lines)})
    # Tin filler — plain list
    tin_filler = layer.get("tin_filler") or []
    if tin_filler:
        filler_lines = "; ".join(
            f"{item.get('tieu_de', '')} [{item.get('tag', '')}]"
            for item in tin_filler
            if isinstance(item, dict)
        )
        fields.append({"label": "Tin filler", "value": parse_fragments(filler_lines)})
    if layer.get("tac_dong"):
        fields.append({"label": "Tác động", "value": parse_fragments(str(layer["tac_dong"]))})
    return fields


_LAYER_FIELD_BUILDERS = {
    "L1": _layer_fields_l1,
    "L2": _layer_fields_l2,
    "L3": _layer_fields_l3,
    "L4": _layer_fields_l4,
    "L5": _layer_fields_l5,
}


def _build_layer_card(layer_key: str, layer: dict[str, Any], *, is_first: bool) -> dict[str, Any]:
    status_label = layer.get("statusLabel", "")
    diff_text = layer.get("diff", "")

    field_builder = _LAYER_FIELD_BUILDERS.get(layer_key)
    fields = field_builder(layer) if field_builder else []

    return {
        "layerNum": layer_key,
        "layerName": _LAYER_NAMES.get(layer_key, layer_key),
        "statusLabel": status_label,
        "statusLevel": status_level(layer_key, status_label),
        "fields": fields,
        "diff": _build_diff(diff_text, is_first=is_first),
    }


# ── Briefing builder ─────────────────────────────────────────────────────────


def _build_briefing(l6: dict[str, Any], *, is_first: bool, updated_at: str) -> dict[str, Any]:
    trend = l6.get("trend", "")
    status = l6.get("status", "")
    timeframe = l6.get("timeframe", "")
    narrative_text = l6.get("narrative", "")
    diff_text = l6.get("diff", "")
    recommendation = l6.get("recommendation", "Quan sát thêm")

    # observations
    obs_raw = l6.get("observations") or {}
    observations = {
        "liquidity": parse_fragments(obs_raw.get("liquidity", "")),
        "moneyFlow": parse_fragments(obs_raw.get("moneyFlow", "")),
        "insider": parse_fragments(obs_raw.get("insider", "")),
        "news": parse_fragments(obs_raw.get("news", "")),
        "supportResistance": parse_fragments(obs_raw.get("supportResistance", "")),
    }

    # watchLevels — pass through as-is (already {tag, description})
    watch_levels = l6.get("watchLevels") or []

    # Guard recommendation to the 5 valid phrases
    if recommendation not in _RECOMMENDATION_SET:
        recommendation = "Quan sát thêm"

    return {
        "updatedAt": updated_at,
        "trend": trend,
        "status": status,
        "statusVariant": briefing_variant(trend, status),
        "timeframe": timeframe,
        "narrative": parse_fragments(narrative_text),
        "diff": _build_diff(diff_text, is_first=is_first),
        "observations": observations,
        "watchLevels": watch_levels,
        "recommendation": recommendation,
    }


# ── Public API ───────────────────────────────────────────────────────────────


def build_insight_response(
    ai_json: dict[str, Any],
    payload: dict[str, Any],
    prev: dict[str, Any] | None,
) -> dict[str, Any]:
    """Assemble the v2 AIInsightResponse dict.

    Args:
        ai_json:  Parsed LLM JSON with keys L1..L6 (per task-1-report.md schema).
        payload:  build_insight_payload() output (price_board, company_overview,
                  company_details, rawInput, ...).
        prev:     Previous analysis dict (or None for first analysis).

    Returns:
        Dict matching SPEC §4.5 AIInsightResponse.
    """
    is_first = prev is None
    updated_at = datetime.now(UTC).isoformat()
    symbol = str(payload.get("symbol", "")).upper()

    # Header
    header = _build_header(payload)
    if not header["symbol"] and symbol:
        header["symbol"] = symbol

    # Layers L1–L5
    layers: dict[str, Any] = {}
    for layer_key in ("L1", "L2", "L3", "L4", "L5"):
        layer_data = ai_json.get(layer_key) or {}
        layers[layer_key] = _build_layer_card(layer_key, layer_data, is_first=is_first)

    # Briefing from L6
    l6 = ai_json.get("L6") or {}
    briefing = _build_briefing(l6, is_first=is_first, updated_at=updated_at)

    # rawInput — carry through from payload (BT8 populates it; here we pass through)
    raw_input = payload.get("rawInput") or {}

    return {
        "symbol": symbol or header["symbol"],
        "updatedAt": updated_at,
        "header": header,
        "briefing": briefing,
        "layers": layers,
        "rawInput": raw_input,
    }

File: backend-v2/src/modules/analysis/insight-response.ts
type Json = Record<string, unknown>;
type Fragment = {
  type: 'text' | 'number' | 'highlight' | 'emphasis';
  content: string;
  variant?: string;
};

const LEVELS: Record<string, Record<string, number>> = {
  L1: { 'Rất yếu': 1, Yếu: 2, 'Trung bình': 3, Mạnh: 4, 'Rất mạnh': 5 },
  L2: { 'Rất yếu': 1, Yếu: 2, 'Bình thường': 3, Mạnh: 4, 'Rất mạnh': 5 },
  L3: { 'Cảnh báo mạnh': 1, 'Cảnh báo nhẹ': 2, 'Trung tính': 3, 'Hỗ trợ nhẹ': 4, 'Hỗ trợ mạnh': 5 },
  L4: { 'Cảnh báo mạnh': 1, 'Cảnh báo nhẹ': 2, 'Trung tính': 3, 'Hỗ trợ nhẹ': 4, 'Hỗ trợ mạnh': 5 },
  L5: { 'Rất tiêu cực': 1, 'Tiêu cực': 2, 'Trung tính': 3, 'Tích cực': 4, 'Rất tích cực': 5 },
};
const NAMES: Record<string, string> = {
  L1: 'Xu hướng',
  L2: 'Thanh khoản',
  L3: 'Dòng tiền',
  L4: 'Nội bộ',
  L5: 'Tin tức',
};
const FIELDS: Record<string, readonly [string, string][]> = {
  L1: [
    ['xu_huong', 'Xu hướng'],
    ['statusLabel', 'Trạng thái'],
    ['ho_tro', 'Hỗ trợ'],
    ['khang_cu', 'Kháng cự'],
    ['da_gia', 'Đà giá'],
  ],
  L2: [
    ['thanh_khoan', 'Thanh khoản'],
    ['cung_cau', 'Cung–Cầu'],
    ['tac_dong', 'Tác động'],
  ],
  L3: [
    ['khoi_ngoai', 'Khối ngoại'],
    ['tu_doanh', 'Tự doanh'],
    ['statusLabel', 'Tác động'],
  ],
  L4: [
    ['noi_bo', 'Nội bộ'],
    ['khoi_luong_tong', 'Khối lượng tổng'],
    ['statusLabel', 'Tác động'],
  ],
  L5: [
    ['tong_quan', 'Tổng quan'],
    ['tac_dong', 'Tác động'],
  ],
};
const TAG = /\[(bull|bear|warn|info|num|gold|hl)\]([\s\S]*?)\[\/\1\]/g;
const RESIDUAL = /\[\/?(?:bull|bear|warn|info|num|gold|hl)\]/g;

/** Parse model markup into the fragment arrays consumed by the stock UI. */
export function parseInsightFragments(value: unknown): Fragment[] {
  if (typeof value !== 'string' || !value) return [];
  const fragments: Fragment[] = [];
  const addPlain = (part: string) => {
    const content = part.replace(RESIDUAL, '');
    if (content) fragments.push({ type: 'text', content });
  };
  let end = 0;
  for (const match of value.matchAll(TAG)) {
    addPlain(value.slice(end, match.index));
    const content = (match[2] ?? '').replace(RESIDUAL, '');
    const tag = match[1];
    if (tag === 'num') fragments.push({ type: 'number', content });
    else if (tag === 'gold' || tag === 'hl') fragments.push({ type: 'highlight', content });
    else fragments.push({ type: 'emphasis', content, variant: tag });
    end = match.index + match[0].length;
  }
  addPlain(value.slice(end));
  return fragments;
}

function object(value: unknown): Json {
  return value && typeof value === 'object' && !Array.isArray(value) ? (value as Json) : {};
}
function rows(value: unknown): Json[] {
  const list = Array.isArray(value) ? value : object(value).data;
  return Array.isArray(list) ? list.map(object).filter((row) => Object.keys(row).length > 0) : [];
}
function pick(row: Json, keys: string[]): unknown {
  for (const key of keys) if (row[key] !== null && row[key] !== undefined) return row[key];
  return null;
}
function rawInput(payload: Json): Json {
  if (Object.keys(object(payload.rawInput)).length) return object(payload.rawInput);
  const board = rows(payload.price_board)[0] ?? {};
  const ohlcv = rows(payload.ohlcv_30).map((row) => ({
    date: pick(row, ['tradingDate', 'date', 'time', 't']),
    open: pick(row, ['open', 'o']),
    high: pick(row, ['high', 'h']),
    low: pick(row, ['low', 'l']),
    close: pick(row, ['close', 'c']),
    volume: pick(row, ['volume', 'v']),
  }));
  const history = rows(payload.trading_history)
    .slice(0, 10)
    .map((row) => ({
      date: pick(row, ['trading_date', 'tradingDate', 'date']),
      totalVolume: pick(row, ['totalVolume', 'total_volume', 'total_buy_trade_volume']),
    }));
  const flow = (value: unknown) =>
    rows(value)
      .slice(0, 10)
      .map((row) => ({
        date: pick(row, ['tradingDate', 'trading_date', 'date']),
        totalNetVolume: pick(row, [
          'totalNetVolume',
          'total_net_volume',
          'netVolume',
          'net_volume',
        ]),
      }));
  return {
    trend: { realtime: Object.keys(board).length ? board : null, ohlcv, computed: {} },
    liquidity: { latest: history[0] ?? null, avg30: null, history },
    moneyFlow: { foreign: flow(payload.foreign_trade), proprietary: flow(payload.proprietary) },
    insider: { transactions: [] },
    news: { items: rows(payload.news), tickerScore: null },
  };
}
function string(value: unknown): string {
  return typeof value === 'string' ? value : '';
}
function diff(value: unknown, first: boolean): Json {
  const content = string(value);
  return {
    text: parseInsightFragments(content),
    hasChange: !first && !!content.trim() && !/ổn định|lần đầu/i.test(content),
    isFirstAnalysis: first,
  };
}
function newsItems(value: unknown, material: boolean): Json[] {
  if (!Array.isArray(value)) return [];
  return value.map(object).map((item) => ({
    title: string(item.tieu_de),
    tag: string(item.tag),
    ...(material && item.tac_dong_ngan ? { subtitle: string(item.tac_dong_ngan) } : {}),
  }));
}
function layerCard(key: string, raw: Json, first: boolean): Json {
  const statusLabel = string(raw.statusLabel);
  const fields = (FIELDS[key] ?? []).flatMap(([source, label]) =>
    raw[source] ? [{ label, value: parseInsightFragments(raw[source]) }] : [],
  );
  if (key === 'L5') {
    const material = newsItems(raw.tin_material, true);
    const filler = newsItems(raw.tin_filler, false);
    if (material.length)
      fields.splice(1, 0, {
        label: 'Tin trọng yếu',
        value: parseInsightFragments(
          material
            .map(
              (item) => `${item.title} [${item.tag}]${item.subtitle ? ` — ${item.subtitle}` : ''}`,
            )
            .join('; '),
        ),
      });
    if (filler.length)
      fields.splice(material.length ? 2 : 1, 0, {
        label: 'Tin phụ',
        value: parseInsightFragments(
          filler.map((item) => `${item.title} [${item.tag}]`).join('; '),
        ),
      });
  }
  return {
    layerNum: key,
    layerName: NAMES[key],
    statusLabel,
    statusLevel: LEVELS[key]?.[statusLabel],
    fields,
    diff: diff(raw.diff, first),
    ...(key === 'L5'
      ? {
          news: {
            material: newsItems(raw.tin_material, true),
            filler: newsItems(raw.tin_filler, false),
          },
        }
      : {}),
  };
}
function briefingVariant(trend: string, status: string): string {
  const level = LEVELS.L1?.[status] ?? 3;
  if (trend.includes('Giảm')) return 'bear';
  if (trend.includes('Tăng') && level >= 4) return 'bull';
  if (trend.includes('Đi ngang') && level <= 2) return 'warn';
  return level <= 2 ? 'bear' : 'neutral';
}

export function buildInsightResponse(
  raw: Json,
  payload: Json,
  header: Json,
  updatedAt: string,
  first: boolean,
): Json {
  const l6 = object(raw.L6);
  const observations = object(l6.observations);
  const trend = string(l6.trend);
  const status = string(l6.status);
  return {
    symbol: payload.symbol,
    updatedAt,
    header,
    briefing: {
      updatedAt,
      trend,
      status,
      statusVariant: briefingVariant(trend, status),
      timeframe: string(l6.timeframe),
      narrative: parseInsightFragments(l6.narrative),
      diff: diff(l6.diff, first),
      observations: Object.fromEntries(
        ['liquidity', 'moneyFlow', 'insider', 'news', 'supportResistance'].map((key) => [
          key,
          parseInsightFragments(observations[key]),
        ]),
      ),
      watchLevels: Array.isArray(l6.watchLevels) ? l6.watchLevels : [],
      recommendation: l6.recommendation,
    },
    layers: Object.fromEntries(
      ['L1', 'L2', 'L3', 'L4', 'L5'].map((key) => [key, layerCard(key, object(raw[key]), first)]),
    ),
    rawInput: rawInput(payload),
  };
}
